sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -639,6 +640,24 @@ object HiveTypeCoercion {
   }
 
   /**
+   * Turns Add/Subtract of DateType/TimestampType and IntervalType to TimeAdd/TimeSub
+   */
+  object DateTimeOperations extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+      // Skip nodes who's children have not been resolved yet.
+      case e if !e.childrenResolved => e
+
+      case Add(left, right) if left.dataType == IntervalType =>
+        Add(right, left) // switch the order
+
+      case Add(left, right) if right.dataType == IntervalType =>
+        Cast(TimeAdd(Cast(left, TimestampType), right), left.dataType)
```

Comment:
this implicitly casts all types to TimestampType. How about requiring left to be string, date, or timestamp?